File: go/pkg/tree/field.go
package tree

import (
	"fmt"

	"github.com/shahzodshafizod/ptaskbook/go/pkg/node"
)

const (
	SPACE  int8 = 0
	LEFT   int8 = 1
	CENTER int8 = 2
	RIGHT  int8 = 3
	DATA   int8 = 4
)

type Field struct {
	nodes     [][]*node.Node
	chars     [][]int8
	height    int
	width     int
	cellWidth int
	left      string
	center    string
	right     string
}

func (f *Field) Init(n *node.Node, height int, width int) {
	f.height = height
	f.width = width
	f.left = string(rune(9484))
	f.center = string(rune(9472))
	f.right = string(rune(9488))
	f.cellWidth = 4
	f.nodes = make([][]*node.Node, f.height)
	f.chars = make([][]int8, f.height)
	for row := 0; row < f.height; row++ {
		f.nodes[row] = make([]*node.Node, f.width)
		f.chars[row] = make([]int8, f.width)
	}
	var column int = 0
	f.fillNodes(n, 0, &column)
	f.fillChars()
}

func (f *Field) fillNodes(node *node.Node, row int, col *int) {
	if node == nil {
		return
	}
	f.fillNodes(node.Left, row+1, col)
	f.nodes[row][*col] = node
	(*col)++
	f.fillNodes(node.Right, row+1, col)
}

func (f *Field) fillChars() {
	for row := 0; row < f.height; row++ {
		for col := 0; col < f.width; col++ {
			if f.nodes[row][col] == nil {
				continue
			}
			f.chars[row][col] = DATA
			if f.nodes[row][col].Left != nil {
				for k := col - 1; k >= 0; k-- {
					if f.nodes[row+1][k] != nil {
						f.chars[row][k] = LEFT
						break
					}
					f.chars[row][k] = CENTER
				}
			}
			if f.nodes[row][col].Right != nil {
				for col++; col < f.width; col++ {
					if f.nodes[row+1][col] != nil {
						f.chars[row][col] = RIGHT
						break
					}
					f.chars[row][col] = CENTER
				}
			}
		}
	}
}

func (f Field) getNextNode(row int, col int) *node.Node {
	for col++; col < f.width; col++ {
		if f.nodes[row][col] != nil {
			return f.nodes[row][col]
		}
	}
	return nil
}

func getLength(number int) int {
	var len int = 0
	if number < 0 {
		len++
		number *= -1
	}
	for number > 0 {
		len++
		number /= 10
	}
	return len
}

func printChars(str string, count int) {
	for count > 0 {
		fmt.Print(str)
		count--
	}
}

func (f Field) Display() {
	var (
		nextNode           *node.Node
		data, spaces, half int
	)
	for row := range f.chars {
		for col := range f.chars[row] {
			switch f.chars[row][col] {
			case SPACE:
				printChars(" ", f.cellWidth)
			case LEFT:
				nextNode = f.getNextNode(row, col)
				data = nextNode.Data
				spaces = f.cellWidth - getLength(data)
				half = spaces / 2
				printChars(" ", half)
				printChars(f.left, 1)
				printChars(f.center, f.cellWidth-1)
			case CENTER:
				printChars(f.center, f.cellWidth)
			case RIGHT:
				printChars(f.center, f.cellWidth-1)
				printChars(f.right, 1)
				printChars(" ", spaces-half)
			case DATA:
				if f.nodes[row][col].Left == nil {
					data = f.nodes[row][col].Data
					spaces = f.cellWidth - getLength(data)
					half = spaces / 2
					printChars(" ", half)
				}
				fmt.Print(data)
				if f.nodes[row][col].Right == nil {
					printChars(" ", spaces-half)
				}
			}
		}
		fmt.Println()
	}
}

File: php/19tree/classes/Field.php
<?php

require_once "Node.php";

class Field {
	private $nodes;
	private $chars;
	private $heigth;
	private $width;
	private $left;
	private $center;
	private $right;
	private $cellWidth;
	public function __construct($node, $height, $width) {
		$this->height = $height;
		$this->width = $width;
		$this->cellWidth = 4;
		$this->left = '<';
		$this->right = '>';
		$this->center = '-';
		$this->init($node);
	}
	private function init($node) {
		for ($i = 0; $i < $this->height; $i++) {
			for ($j = 0; $j < $this->width; $j++) {
				$this->nodes[$i][$j] = null;
				$this->chars[$i][$j] = SPACE;
			}
		}
		$col = 0;
		$this->fillNodes($node, 0, $col);
		$this->fillChars();
	}
	private function fillNodes($node, $row, &$col) {
		if ($node == null) {
			return;
		}
		$this->fillNodes($node->left, $row+1, $col);
		$this->nodes[$row][$col++] = $node;
		$this->fillNodes($node->right, $row+1, $col);
	}
	private function fillChars() {
		for ($i = 0; $i < $this->height; $i++) {
			for ($j = 0; $j < $this->width; $j++) {
				if ($this->nodes[$i][$j] == null) {
					continue;
				}
				$this->chars[$i][$j] = DATA;
				if ($this->nodes[$i][$j]->left != null) {
					for ($k = $j-1; $k >= 0; $k--) {
						if ($this->nodes[$i+1][$k] != null) {
							$this->chars[$i][$k] = LEFT;
							break;
						}
						$this->chars[$i][$k] = CENTER;
					}
				}
				if ($this->nodes[$i][$j]->right != null) {
					for ($j++; $j < $this->width; $j++) {
						if ($this->nodes[$i+1][$j] != null) {
							$this->chars[$j][$j] = RIGHT;
						}
						$this->chars[$i][$j] = CENTER;
					}
				}
			}
		}
	}
	public function display() {
		for ($i = 0; $i < $this->height; $i++) {
			$nextNode = null;
			$data = 0;
			$spaces = 0;
			$half = 0;
			for ($j = 0; $j < $this->width; $j++) {
				switch ($this->chars[$i][$j]) {
					case SPACE:
						$this->printChars(' ', $this->cellWidth);
						break;
					case LEFT:
						$nextNode = $this->getNode($i, $j);
						$data = $nextNode->data;
						$spaces = $this->cellWidth - $this->getLength($data);
						$half = intval($spaces / 2);
						$this->printChars(' ', $half);
						$this->printChars($this->left, 1);
						$this->printChars($this->center, $this->cellWidth - 1);
						break;
					case CENTER:
						$this->printChars($this->center, $this->cellWidth);
						break;
					case RIGHT:
						$this->printChars($this->center, $this->cellWidth - 1);
						$this->printChars($this->right, 1);
						$this->printChars(' ', $spaces - $half);
						break;
					case DATA:
						if ($this->nodes[$i][$j]->left == null) {
							$nextNode = $this->nodes[$i][$j];
							$data = $nextNode->data;
							$spaces = $this->cellWidth - $this->getLength($data);
							$half = intval($spaces / 2);
							$this->printChars(' ', $half);
						}
						echo $data;
						if ($this->nodes[$i][$j]->right == null) {
							$this->printChars(' ', $spaces - $half);
						}
						break;
				}
			}
			echo "\n";
		}
	}
	private function printChars($c, $count) {
		for ($i = 0; $i < $count; $i++) {
			echo $c;
		}
	}
	private function getLength($data) {
		$len = 0;
		if ($data < 0) {
			$len++;
			$data *= -1;
		}
		while ($data > 0) {
			$len++;
			$data = intval($data / 10);
		}
		return $len;
	}
	private function getNode($row, $col) {
		for ($col++; $col < $this->width; $col++) {
			if ($this->nodes[$row][$col] != null) {
				return $this->nodes[$row][$col];
			}
		}
		return null;
	}
}

?>